docs/qos_enforcement.md

```diff
@@ -0,0 +1,143 @@
+# Magma QoS Enforcement
+
+*Status: In-Review*\
+*Author: @andreilee*\
+*Last Updated: 07/28*
+
+This document concerns:
+
+1. The addition of configurable QoS
+2. The addition of more comprehensive associations between subscribers, QoS, policies, and APN
+3. The enforcement of the above with an eventually consistent model
+
+
+## What We Want to Accomplish
+
+We would like to enable QoS enforcement, and move towards a model that more closely enforces across UE level, APN level, and bearer level.
+
+As an example use case for QoS enforcement, a service provider may wish to segment users into tiers denoting their service quality, eg. bronze, silver, and gold. A service provider should then be able configure a QoS profiles for each of the three service levels. There should be the capability to define which APNs support which QoS profiles, and which subscribers are allowed which QoS levels, based on their subscriptions. Naturally, different policy rules will also be tied to different QoS profiles, and together with APN restrictions on supported QoS profiles, this places a restriction on which policy rules may be installed for different users.
+
+As an addendum to this, we wish to enable QoS enforcement for multiple sessions of a single subscriber, one for each APN. Across these different sessions, each APN will have a different QoS for its default bearer. To enable this behavior, there should be the capability to assign a QoS profile to an APN for the default bearer.
+
+To enable the above, we have the following list of new API requirements for Magma:
+
+1. Configuration of QoS profiles
+2. Configuration of APN entities
+3. Assignment of QoS profiles to subscribers (to denote what QoS levels are allowed for the subscriber)
+4. Assignment of QoS profiles to APN entities (to denote what QoS levels are supported by the APN)
+5. Assignment of APN entities to subscribers (to denote which APNs a subscriber has access to)
+
+## How We Will Change Magma
+
+### API Endpoints
+
+To enable configuration of QoS profiles and APNs, we will support the following REST endpoints:
+
+```
+**QoS Profiles
+
+GET     ****/****lte****/{****network_id****}/****qos_profiles**
+**POST    ****/****lte****/{****network_id****}/****qos_profiles**
+
+**GET     /lte/{network_id}/qos_profiles/{qos_name}
+PUT     /lte/{network_id}/qos_profiles/{qos_name}
+DELETE  /lte/{network_id}/qos_profiles/{qos_name}
+
+
+APN Profiles**
+
+**GET     ****/****lte****/****{****network_id****}/****apns**
+**POST    ****/****lte****/{****network_id****}/****apns**
+
+**GET     ****/****lte****/{****network_id****}/****apns****/{****apn_name****}**
+**PUT     ****/****lte****/{****network_id****}/****apns****/{****apn_name****}**
+**DELETE  ****/****lte****/{****network_id****}/****apns****/{****apn_name****}**
+
+
+```
+
+
+
+### API Entity Definitions
+
+Entity definitions for QoS, APN, and subscribers will change. The final state is summarized here:
```

Comment:
Let's add back the GBR field. This will make it consistent with the spec. Even if AGW is not enforcing it, RAN can still enforce it (particularly important for the wireless scheduler).

Reply:
also note that PCRF also utilizes these fields...

Reply:
Sure I'll add it back in